kernel/kernbench.py

```diff
@@ -0,0 +1,161 @@
+from avocado.utils import cpu
+from avocado.utils import distro
+from avocado.utils.software_manager import SoftwareManager
+
+log = logging.getLogger('avocado.test')
+
+
+class Kernbench(Test):
+    """
+    Kernbench compiles the kernel source fetched from the kernel.org
+    and compile it as per the configuration file. Performance figures will be
+    shown in the log file.
+    """
+
+    def time_build(self, kern,
+                   threads=None, timefile=None, make_opts=None):
+        """
+        Time the bulding of the kernel
+        """
+        os.chdir(kern.linux_dir)
+        build.make(kern.linux_dir, extra_args='clean')
+        if kern.config_path is None:
+            build.make(kern.linux_dir, extra_args='defconfig')
+        else:
+            build.make(kern.linux_dir, extra_args='oldnoconfig')
+        if make_opts:
+            build_string = "/usr/bin/time -o %s make %s -j %s vmlinux" % (
+                timefile, make_opts, threads)
+        else:
+            build_string = "/usr/bin/time -o %s make -j %s vmlinux" % (
+                timefile, threads)
+        process.system(build_string, allow_output_check='None', shell=True)
+        if not os.path.isfile('vmlinux'):
+            self.error("No vmlinux found, kernel build failed")
+
+    @staticmethod
+    def to_seconds(time_string):
+        """
+        Converts a string in M+:SS.SS format to S+.SS
+        """
+        elts = time_string.split(':')
+        if len(elts) == 1:
+            return time_string
+        return str(int(elts[0]) * 60 + float(elts[1]))
+
+    def extract_all_time_results(self, results_string):
+        """
+        Extract user, system, and elapsed times into a list of tuples
+        """
+        pattern = re.compile(r"(.*?)user (.*?)system (.*?)elapsed")
+        results = []
+        for result in pattern.findall(results_string):
+            results.append(tuple([self.to_seconds(elt) for elt in result]))
+        return results
+
+    def setUp(self):
+        """
+        Setting up the env for the kernel building
+        """
+        smg = SoftwareManager()
+        detected_distro = distro.detect()
+        deps = ['gcc', 'make', 'automake', 'autoconf', 'time']
+        if 'Ubuntu' in detected_distro.name:
+            deps.extend(['libpopt0', 'libc6', 'libc6-dev', 'libpopt-dev',
+                         'libcap-ng0', 'libcap-ng-dev', 'elfutils', 'libelf1',
+                         'libnuma-dev', 'libfuse-dev'])
+        elif 'SuSE' in detected_distro.name:
+            deps.extend(['libpopt0', 'glibc', 'glibc-devel',
+                         'popt-devel', 'libcap1', 'libcap-devel',
+                         'libcap-ng-devel'])
+        # FIXME: "redhat" as the distro name for RHEL is deprecated
+        # on Avocado versions >= 50.0.  This is a temporary compatibility
+        # enabler for older runners, but should be removed soon
+        elif detected_distro.name in ['centos', 'fedora', 'rhel', 'redhat']:
```

Comment:
i think we can  drop redhat as we are using avocado versions >= 50.0 what you say ?

Reply:
missed that. will remove `redhat`.